### ISSUE TITLE
DFDL Specification review comments

### ISSUE THREAD

**Issue by brutzman** [OpenGridForum/Editor#16]
Thanks for many stellar efforts on DFDL specification and implementations.

Attached please find DFDL Specification review comments.  If you prefer these can be separated into individual issues.

https://gitlab.nps.edu/Savage/robodata/-/blob/master/DFDL/pipeline/SpecificationCommentsBrutzmanDecember2020.txt

---

Technical.

1. Section 1.3 "What DFDL is not" states "cannot use XML attributes in the
data model."

This is a significant deficiency in DFDL that greatly inhibits adoption.
This functional gap means that many existing schemas (both standardized and
informal) cannot be used directly with DFDL.

It is also hard to believe that another version of DFDL is advancing without
addressing this significant and longstanding shortfall.

Given that DFDL has successfully accomplished so much important functionality
across such a diverse range of formats, the ability to define XML attributes
as well as XML child elements certainly seems comparable to other existing
DFDL capabilities.

Am attempting a workaround that automatically modifies XML documents and
corresponding XSD schemas and as part of a pipeline for DFDL preprocessing
and postprocessing.  Preliminary results are positive.
https://gitlab.nps.edu/Savage/robodata/-/blob/master/DFDL/pipeline

---

2. How are document comments handled?  Presumably an author writes DFDL to parse
them according to whatever syntax is appropriate, then provides a named object
container for them in XML.  However, comments are often referred to as either
required to be ignored or optionally retained in other standards.  Recommend
adding guidance that handling comments is allowed but can be handled like any
other data item.

---

3. Paragraph 5.3.5 Enumeration p.24 of 246

a. Change title "Enumeration" to "Enumerations" or "Enumeration Values"

b. Current prose (surprisingly) says:

"Enumerations are used to provide a list of valid values in XSD.
• Used for validation only
Note: in DFDL we do not use XSD enumeration as a means to define symbolic constants. These
are captured using dfdl:defineVariable constructs so they can be referenced from expressions."

Enumerations are a core part of any data model and are further are also part of
metadata models.  Such definitions can become quite lengthy and often include
annotations for semantic definitions and reference urls.

Duplicating such information is similarly verbose and an error-prone process.
Shouldn't there be a way to tersely apply dfdl: markup to annotated enumerations,
avoiding problems associated with duplication?  This would be more in keeping
with other DFDL constructs that complement already-existing XSD structures.

---

Editorial.

1. Major.  Specification version number is confusing.  "1.0" should not be
re-used since the syntax and semantics are different that prior versions
(otherwise another specification document like this one is not needed).
Overloading version "1.0" makes referring to the correct document quite
error prone, it is trivially easy to use the wrong document.

If changes are considered minor relative to overall design of original DFDL,
then indicate change in the minor version number: version "1.1" or version "1.2"
instead.

Distinctly relabeling the specification does not necessarily have to change the
XML Namespace for Daffodil, i.e.
    xmlns:dfdl="http://www.ogf.org/dfdl/dfdl-1.0/"
can remain unchanged. The updated specification might also equivalently allow
    xmlns:dfdl="http://www.ogf.org/dfdl/dfdl-1.1/"

2. Fix bookmarks in HTML version to be human readable, rather than autogenerated
numbers.  This make document references much more repeatable and less error prone,
also adds to forward compatibility for referencing sections in future versions.

3. What is status of GFD.190, "Mapping between DFDL 1.0 Infoset and XML Data Model"
which is not mentioned in the specification but is listed as relevant on website
in first paragraph at https://www.ogf.org/ogf/doku.php/standards/dfdl/dfdl44

4. Introduction section 1, pp 8-9 of 246.

a. Omit "very" from "very high-performance data handling" since it adds no
particular meaning.  This is a general comment.

b. Precision: change "2. Binary data in standard formats." to
"2. Binary data in regular formats."

c. Third person: multiple occurrences of "we" are improper for a technical
reference, as a scientific/engineering standard, all such occurrences of "we"
need to be in third person.

Similarly occurrences of second person "you" also need to be rephrased (for
example in section 1.3 page 12).

d. Footnote 1 p. 8, insert space characters in "Rules(CER)[ASN1CER]" as
"Rules (CER) [ASN1CER]".

e. Precision: insert "regular" in "almost any possible representation scheme"
as "almost any possible regular representation scheme".

5. Section 1.1, p 9 of 246.

a. Avoid overstatement: insert "often" in "using an XML-based data format is harder"
as "using an XML-based data format is often harder".

b. Grammatical: insert hyphen in compound adjective, modifying
"smoothly integrate standardized data format handling" as
"smoothly integrate standardized data-format handling".

c. Clarity: insert "evolving" in "where a program first becomes important" as
"where an evolving program first becomes important".

d. Clarity: insert "evolving" in "after-the-fact description of data formats" as
"after-the-fact description of evolving data formats".

e. Footnote 3, clarity:  insert "terms" in "but those both connote a sequencing order"
as "but those terms both connote a sequencing order".

6. Footnotes 4 and 5, p. 19 of 246, both refer to Post-Schema Validation
Infoset (PSVI) without providing a reference.  This reference always seems
to be tricky to find, I think there is a note and (at least) a mention
in XML Schema Recommendations.  For thoroughness probably should also refer
to [XMLInfoset].

7. 1.2.1 Simple Example p. 10 of 246:

a. insert "a" in
"document that simply provides description" as
"document that simply provides a description"

b. Editorial, avoid use of "could" since it is ambiguous - this is a global
comment.  For example, change
"A binary representation of the data could be visualized" to
"A binary representation of the data can be visualized".

c. Precision, avoiding indirect phrasing: change "To describe this in DFDL," to
"To describe the same information in DFDL,"

8. Section 1.4 Scope, p. 13 of 246.

a. General features. Change
"g) Basic math - in DFDL expressions" to
"g) Basic arithmetic in DFDL expressions." (no hyphen, append period)

b. Paragraph 5, "HL7" is a duplicate entry.

9. 4.3 DFDL Information Item Order, p. 19 of 246.
"On parsing and unparsing information items will be presented in the order
they are defined in the DFDL Schema."

Stay in present tense throughout, i.e. "are presented in the order that".
This is a global editorial comment.

10. Paragraph 5.1 DFDL Simple Typess p. 21 of 246,
Fix punctuation, change
"None are reserved. (Including the character..." to
"None are reserved (including the character..."

11. Paragraph 6.1 Namespaces, p. 26 of 246. For single-sentence paragraph
"Attributes on DFDL annotations that are not in the DFDL namespace or
in no namespace are ignored"
append "by a DFDL processor".

12. Paragraph 6.3 DFDL Properties, awkward phrasing:
"A DFDL property is a specific DFDL syntax that tells the DFDL processor
something about the data format."

This sentence can be improved, perhaps something like:
"A DFDL property is a specific DFDL construct that tells the DFDL processor
some characteristic about the data format.

13. Paragraph 9.6 Validation, p. 73 of 246: insert "DFDL" as follows:
"This implies that Validation Errors cannot affect the parsing or unparsing of data." to
"This implies that DFDL Validation Errors cannot affect the parsing or unparsing of data."

Further, I'd question the inclusion of "or unparsing" in this sentence, since
if an error is present that might well prevent unparsing from occurring at all.

14. Paragraph 18.3 General Syntax, p. 178 of 246. Apparent omission:
'When a property accepts either a DFDL string literal or a DFDL expression, and the value is a
string literal starting with a "{" character, then "{{"must be used to escape the "{" character.'

Presumably need to add 'and "}}"must be used to escape the "}" character.'

15. 26 Intellectual Property Statement

a. Need link to OGF IPR policy governing specification contributions and development.

b. Need link to online declaration of known claims (if any).

16. Paragraph 31 Appendix B: Rationale for Single-Assignment Variables p 208.

a. Might note that this approach is a form of immutable objects as defined
in other programming languages.

* Wikipedia: Immutable object
  https://en.wikipedia.org/wiki/Immutable_object

17 Glossary, "Content Model" p. 219 of 246: note "and not mixed content."

* Extensible Markup Language (XML) 1.0 (Fifth Edition), 3.2.2 Mixed Content
  https://www.w3.org/TR/xml/#sec-mixed-content

18 Glossary, "Schema Component Designator (SCD)" p. 222 of 246 states
"A notation for referring to one of the components of a
DFDL Schema. This is being standardized by W3C. See http://www.w3.org/TR/xmlschema-ref."

* 3.1 Schema Component Designator Syntax
  https://www.w3.org/TR/xmlschema-ref/#section-scd-syntax

Status remains W3C Candidate Recommendation 19 January 2010, suggest omitting
or else confirming active status.

**Comment by mbeckerle** [comment]
Thank you for these excellent comments and thorough review.

Several things you mention above need responses here. Everything else (from Editorial item #4) down are all things I think we will incorporate and/or provide clarifications about. 

The things that need specific comments (which will follow this comment) are:
* attributes
* comments
* enumerations
* specification version number
* HTML bookmarks
* status of GFD.190, "Mapping between DFDL 1.0 Infoset and XML Data Model"

**Comment by mbeckerle** [comment]
re: Attributes

DFDL chose to use XSD as a platform to get "a popular standard BNF with a standard annotation syntax". 

Attribute declarations in XSD actually throw a big wrench into using XSD as "a standard BNF". Attribute declarations must be provided in a specific location (after all child-element declarations) so one cannot rely on schema-declaration order the way DFDL does now for element declarations. This and the fact that DFDL doesn't need attributes in order to express any data formats are why this wasn't pursued for DFDL v1.0. 

There is the degenerate case where all children are simple type, and so all children can be attributes. A concrete proposal for using attributes or even just this degenerate sub-case would certainly be entertained for DFDL v2.0. Keep in mind that the DFDL infoset created by parsing, consumed by unparsing, must also be something directly projected into JSON, C structs, and other popular data formats. Most non-XML data formats do not have the dual-namespaces for element children and attribute children that XML allows, so this is another reason we bypassed using XSD attributes in DFDL. 

The need for attributes for XML users has made it to a [Daffodil-implementation wish-list page](https://cwiki.apache.org/confluence/display/DAFFODIL/DFDL+Wish+List).

**Comment by mbeckerle** [comment]
re: comments

Your point concluded 

    "Recommend adding guidance that handling comments is allowed but can be handled like any other data item."

I am not sure what this means. My basic understanding is when unparsing, incoming XML must be an XML representation corresponding to the DFDL Infoset. The DFDL Infoset has no notion of comments; hence, XML comments would have to be dropped.  When parsing data XML Comments would never be generated in the XML created from the DFDL Infoset. 

Is there some other role you envision for XML comments?

**Comment by mbeckerle** [comment]
re: enumerations

Agreed that this is an important feature. There are prototypes that enable conversion of integer values to/from corresponding enumeration constants. We're not happy with them (too complex) and plan to evolve (simplify) the prototypes and gain experience before advocating incorporation into DFDL; hence, this issue did not make the cut for DFDL v1.0. 

This issue is open for DFDL v2.0 here: https://github.com/OpenGridForum/DFDL/issues/14

**Comment by mbeckerle** [comment]
re: Specification Version Number

The challenge here has been OGF official documents, vs. working group drafts that aren't yet official. Plus even the term "draft" has specific formal meaning.  The idea here is DFDL v1.0 is a language, and various spec drafts are increasingly accurate descriptions of it. So there really are two distinct version spaces.  I think we need to just split the two, so 

    DFDL v1.0 rev 2021-01-07

Is one possible scheme taking advantage of publication date as the document version.

**Comment by mbeckerle** [comment]
re: HTML Bookmarks

The HTML version, while useful, is unofficial.  
For this version of the spec., we are stuck with tooling, MS-Word, which doesn't do a great job of generating HTML. 

My hope, in the future, is to convert this MS-Word to an XML-representation of the DFDL spec., so that it is machine-readable, and an XML-pipeline can turn it into HTML, PDF, etc. This will also allow more sensible change control deltas by treating the XML like source code, instead of a giant MS-Word "blob".

**Comment by alansill** [comment]
Re. OGF document revisions: All OGF documents have a distinct Grid Forum Document (GFD) number, which should be used as the definitive reference to identify the document. The process was designed to support even lengthy initial construction of a document by keeping its status as a Grid Working Draft (GWD) with various qualifications possible to this designator as the document matured. The process was not designed to allow continuous revision of a published specification except for errata once it has been published. As spelled out in [GFD.152](https://www.ogf.org/documents/GFD.152.pdf): 

> Once approved for publication as a OGF document, the document will be changed by the OGF
> Editor as follows:
> 
> - The document type will be updated from GWD to GFD and a GFD sequence number assigned.
> - The Status field will be updated from “Group working Draft (GWD)” to “Current.”
> - The Document date will be set to the date of first publication; any prior dates will be removed.
> - The revision date will be set to the date(s) of any change allowed here after first publication (described below under Errata). The document change history field will briefly describe the changes. 

The Errata section describes the process for minor un-versioned fixes, and a nominal 6-month cutoff for substantial changes. It was not envisioned to allow continuous updates or extensions to a published specification, but a process of obsoleting and replacing specifications with a new GFD was envisioned to accommodate especially breaking changes to a published document and replacement with a new specification. 

Again from GFD.152:

> Whenever a document is updated, even for minor editorial updates, the document header will be
> adjusted to reflect the date the document was updated. The date of the update should appear in
> the upper right side of the document's first page, beneath any prior date. For very minor updates,
> especially those within the first month after publication, no additional information needs to be
> added to the document.
> 
> For all other updates, minor and major technical fixes, an errata report must be added to the
> document itself. Recommended practice is to put an errata report labeled as “Document Change
> History” on the first page of the document, directly under the copyright statement and before the
> abstract. A brief report on what was changed is sufficient, along with the date. If a more
> discussion is desired, the document change history field (mentioned above) can refer to a later
> section where the update is discussed in detail.
> 
> This errata process is not intended to be a mechanism for obsoleting prior versions of a
> document. If substantial updates are needed, a new document should be created and brought
> through the complete editorial review process. When a new document obsoletes an old
> document, that old document will be edited (header only) to indicate it has been obsoleted by a
> new document. 

So the process does encompass including the date of last revision to specify uniqueness.

With all this in mind, the recommendation for the process boils down to using working draft (GWD) document specifiers while a spec is being developed, publishing it as a final Grid Forum Document (GFD) once development is complete and the specification is considered solid and has undergone the requisite public comment and review, and then using errata or the document replacement process to publish a new, substantially changed or incompatible version of the document itself.

This having been said, I am aware of other SDOs that have established patterns of publishing regular updates to documents in which the revision number is promoted to a first-class qualifier for the document, and the document number does not change. An example would be [DMTF](https://www.dmtf.org/standards/published_documents), which also assigns document numbers and revisions to schema files and associated supporting files, allowing them to be updated and versioned separately from the main specifications themselves. 

Feedback as to whether such a versioning pattern would be useful for the DFDL document development process is welcome. I do note a high rate of minor changes that would be hard to track in fixed specification documents.

**Comment by mbeckerle** [comment]
re: status of GFD.190, "Mapping between DFDL 1.0 Infoset and XML Data Model"

This document exists, dated June 2011. From a quick skim, it all seems to still be relevant and correct excepting we have added a small number of things to the DFDL Infoset model such as  [array], [nilled], and [unionMemberSchema] members. 

Implementations can provide details about their mapping to/from XML. (E.g., Daffodil has this page: 
https://daffodil.apache.org/infoset/.)

**Comment by brutzman** [comment]
Thanks for detailed background on process and close scrutiny of versioning.

If I'm tracking your response satisfactorily, seems clear the ability to have a unique identifier is essential.  Here are a few more thoughts on the topic.

Clarity and consistency among adopters and users is of course the chief reason to have precise version identifiability.

Of possible interest, here are a few more use cases where distinct DFDL versioning may be fundamentally important.  These are general goodnesses.

- metadata for schemas, documents and datasets providing archival provenance,
- formal cross-referencing among complementary adopting standards;
- government and industry contract requirements for software and data support,
- detailed unit-test applicability within test suites (which might vary among DFDL versions),
- Semantic Web and Linked Data reasoning on formal terms of reference.

A more precise implementation-oriented use case: for an XML schema containing DFD markup, it would be good to have version with well-defined enumeration values as a required dfdl attribute, so that tools processing the schema have confidence regarding past/present/future version being used.

Typically such information goes in the governing URI or else as an amplifying attribute, which can be helpful if future specification versions support backwards compatibility with prior specification versions.

Along those lines, when looking at public example:

* https://github.com/DFDLSchemas/CSV/blob/master/src/main/resources/com/tresys/csv/xsd/csv.dfdl.xsd

one finds a top-level namespace:

	xmlns:dfdl="http://www.ogf.org/dfdl/dfdl-1.0/"

and an xs:include statement plus dfld:format defaults:

   <xs:include schemaLocation="org/apache/daffodil/xsd/DFDLGeneralFormat.dfdl.xsd" />

   <xs:annotation>
     <xs:appinfo source="http://www.ogf.org/dfdl/">
       <dfdl:format ref="ex:GeneralFormatPortable" separator="" initiator=""
         terminator="" textTrimKind="none" initiatedContent="no" ignoreCase="no"
         separatorPosition="infix" occursCountKind="implicit"
         emptyValueDelimiterPolicy="both" representation="text" textNumberRep="standard"
         lengthKind="delimited" encoding="ASCII" encodingErrorPolicy="error" />
     </xs:appinfo>
   </xs:annotation>

Since the only identifying substring "dfdl-1.0" appears to be identically common to all versions of DFDL specifications, then am not seeing versioning information anywhere.

Just rechecked draft DFDL spec, not finding a definition for versioning anywhere other than 1.0, please advise if I'm missing it.

Looking ahead towards future of DFDL: Section "1.4 Scope of version 1.0" does indeed include a list future features that are not yet supported.  So, without version information, any current DFDL implementation that might fully implement the current specification would apparently be unable to identify a gap if future-version DFDL markup were encountered that included future features.  Instead, today's implementation might only be able to report something like "erroneous unrecognized markup encountered" instead of a more-helpful message noting version mismatch, such as "this implementation does not support DFDL 1.x"

Minor note: date strings can be used for versioning, but such recorded values tend to be error-prone in practice and not very human readable.

**Comment by mbeckerle** [comment]
@brutzman How is this versioning problem specific to DFDL vs. say, the same as what one has for just XML Schema?

E.g., XML Schema has version 1.0 and version 1.1. The namespace is the same. So they're not using the namespace to capture version requirements.

XML Schema 1.1 does define a non-normative URI pattern like http://www.w3.org/XML/XMLSchema/vX.Y/Ne/yyyymmdd by which URIs can be constructed to refer to specific editions of specific versions of the specification. But it doesn't describe any usage idiom for these (from what I can find). 

I guess I'm seeking a concrete proposal here for what would address your needs.

**Comment by smhdfdl** [comment]
@brutzman - many thanks for taking the time to review the spec and for providing comments.

Re - attributes

It was a very deliberate design decision to exclude attributes. @mbeckerle has mentioned some reasons, but it was also influenced by IBM's experience with a proprietary forerunner of DFDL, called MRM, which also used annotated XML Schema and which allowed attributes and most other XML Schema constructs. MRM users whose starting point was a text/binary format to model found it difficult and confusing to use as a result. The decision was made in DFDL 1.0 to drop all XML schema constructs (not just attributes) that are not necessary to model non-XML data, resulting in a cleaner, simpler logical model. The downside of course is that DFDL 1.0 users whose starting point is an existing XML schema are unable to use many such schema without modification. Our experience over the past 10 years of DFDL 1.0 implementations in production is that keeping the logical model simple has been a major factor in its adoption, and outweighs the downsides.

**Comment by brutzman** [comment]
As context/example as to why discernable version numbers have value, XML Schema instances do include a version number for individual schemas.
- W3C XML Schema Definition Language (XSD) 1.1 Part 1: Structures, W3C Recommendation 5 April 2012
- 3.17.2 XML Representations of Schemas
- https://www.w3.org/TR/xmlschema11-1/#declare-schema

> XML Representation Summary: schema Element Information Item et al.

> <schema (listing of attributes)
>  version = token

> "attributes (id and version) are for user convenience, and this specification defines no semantics for them."

**Comment by brutzman** [comment]
Regarding original DFDL specification design decision to avoid representation of XML schema attributes in DFDL, no challenges there.  It is excellent that so much capability has emerged.

Regarding upside/downside, am hoping that after 10 years experience and open-source adoption progress, this restriction might be addressable.  Goal is DFDL might decorate and unlock existing/future XML schemas that include attributes.  In my use cases of current interest, am looking to read alternate encodings or alternate data formats that contain the same/similar data occurring in the original schema.  This might expand usefulness of DFDL for data parsing and correlation.

As linked in topic 1 of this issue, have managed to implement DFDL preprocessing/postprocessing via XSLT that appears to accomplish this goal, testing continues. 

- https://gitlab.nps.edu/Savage/robodata/-/tree/master/DFDL/pipeline
- https://gitlab.nps.edu/Savage/robodata/-/raw/master/DFDL/pipeline/images/DfdlXmlElementAttributeTransformations.png

Approach:

a. Initial XML schema (including dfdl: markup) plus example XML document that can have attributes and elements;
b. XSLT produces corresponding XML schema (including dfdl: markup) and example document, intermediate form has turned each original attribute into an element;
c. DFDL implementation can parse new data instances to produce element-only XML documents;
d. Intermediate-form elements-only XML form then converted via XSLT to original XML form with elements and attributes.

This shows that XML schemas which include attributes might be used satisfactorily with current DFDL.

Given this proof of capability, and given that the XSLT processing is much simpler than what DFDL already accomplishes, further consideration of the original DFDL design decision to disallow attributes in original XML now seems warranted.

**Comment by brutzman** [comment]
@mbeckerle wrote

> re: comments
> 
> Your point concluded
> 
> ```
> "Recommend adding guidance that handling comments is allowed but can be handled like any other data item."
> ```
> 
> I am not sure what this means. My basic understanding is when unparsing, incoming XML must be an XML representation corresponding to the DFDL Infoset. The DFDL Infoset has no notion of comments; hence, XML comments would have to be dropped. When parsing data XML Comments would never be generated in the XML created from the DFDL Infoset.
> 
> Is there some other role you envision for XML comments?

Yes this issue can be ambiguous, will try to explain better.

Your unparse/parse example helps illustrate the problem.  Generally speaking, comments can have value and stripping them is not always desirable, especially if lossless round-trip processing of documents is expected.  However, if a given infoset has no representation for them, then comments simply get dropped and a literal approach to lossless round-trip processing (that restores an original data file with comments) appears impossible.  Furthermore, intermediate XML processors are often neither required nor forbidden (nor configurable) to retain XML comments in XML documents they handle.

Here is hopefully a better suggested prose addition that simply advises DFDL designers that they are "in charge" of comment handling:

> "Comments might allowably occur within data files and corresponding XML files.  If lossless representations and unparse/parse conversions for comments is desired for a given data model, then such persistent comments can be represented by distinct elements in the DFDL schema and XML documents."

Impact: this kind of explanatory statement adds no new requirements to DFDL, and (if we get the words right) might help by explicitly explaining DFDL capabilities with respect to comments.  Remaining silent and implicit can lead some to think that comments are simply not supported in any fashion by DFDL.

Alternative future solution might be to represent comments formally as some kind of explicit datatype in DFDL Infoset.  Feasible but presumably not a desirable task at this stage. 

- Incidentally, have heard some thunderous conversations in other places debating the transient value and ephemeral nature of comments, but if one offers to helpfully remove all comments from every instance of everyone's source code and documents, forever and ever, then participants usually settle down again... 8)

- Curiously, github labels all entries in this issue thread as "comments" and thus yes, comments can have value.

**Comment by mbeckerle** [comment]
@brutzman gotcha. We basically advise people that XML comments are removed, so data formats containing comments must model them in elements, not as XML comments.  Your wording is better.  But I understand your point which is that XML-oriented users may not get this point implicitly and may be looking for a way to map data to XML comments or vice versa when there is no such mechanism.   I will find a place to incorporate this in the section where we discuss XML.

**Comment by mbeckerle** [comment]
I have edited the changes discussed above into the specification, save these:

I have not done anything about versioning. Now that we have version 1.0 of DFDL well specified, we can insure all future documents contain specific version information both on the printed/viewed page, and in the file names, including the HTML version.
 
Point 4b suggest use of the term "regular" vs. standard. This changes the meaning. What is intended there really is dejure or defacto standards. 

Point 4e suggests use of the term regular as in "regular representation". Not changed as this can be confused with the formal use of the term regular as in regular language, regular expression, etc. We tend to prefer "typical" as even the term "normal" has formal meaning (as in normalization).  In this specific sentence, I left it as is, as I did not see that any additional modifier added precision or clarity.
